### PR TITLE
Persian Calendar on Glance

### PR DESCRIPTION
**Pull request by yasanglass**
Adds Persian Calendar support to Glance + A system to manually reload the date content on Glance as it was not previously possible.

The changes also allow other contributors to add their own local calendar system to the project easily.

The Persian Date is always shown in Farsi for now but I will create an additional PR to make the Persian Date respect the language of the device later on as the library used for the Persian Date does not support this feature (yet).

### CODE REVIEW

**Comment by yasanglass** [comment]
Sorry for the force pushes, just a few small adjustments.

**Comment by yasanglass** [comment]
- Improved how `IcuDateTextView.kt` listens to calendar changes.
- Smartspace custom calendar is now experimental & the default value is used when the experimental feature is disabled.
- Default calendar system for Farsi (Iran) is set to Persian in resources.

@paphonb should be ok now, please check again.

**Review by suphon-t — APPROVED** [review]

File: lawnchair/src/app/lawnchair/smartspace/model/SmartspaceCalendar.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+open class SmartspaceCalendar(@StringRes val nameResourceId: Int) {
+
+    companion object {
+        
+        fun fromString(value: String): SmartspaceCalendar {
+            return when (value) {
+                "persian" -> Persian
+                else -> Gregorian
+            }
+        }
```
Comment (ghost) [inline lawnchair/src/app/lawnchair/smartspace/model/SmartspaceCalendar.kt]: An expression body could be used here for conciseness.

```suggestion
        fun fromString(value: String): SmartspaceCalendar =  when (value) {
            "persian" -> Persian
            else -> Gregorian
        }
```

File: AndroidManifest-common.xml [hunk 2]
```diff
@@ -78,7 +78,8 @@
         android:label="@string/derived_app_name"
         android:largeHeap="@bool/config_largeHeap"
         android:restoreAnyVersion="true"
-        android:supportsRtl="true" >
+        android:supportsRtl="true"
+        tools:replace="android:label">
```
Comment (ghost) [inline AndroidManifest-common.xml:82]: Just to confirm, is the addition of `tools:replace="android:label"` intentional? I’m unfamiliar with this attribute.

File: lawnchair/src/app/lawnchair/smartspace/IcuDateTextView.kt [hunk 3]
```diff
@@ -39,27 +51,62 @@ class IcuDateTextView @JvmOverloads constructor(
...
+    private fun getTimeTextPersian(updateFormatter: Boolean): String {
+        if (formatterPersian == null || updateFormatter) {
+            val formatter = PersianDateFormat(
+                context.getString(R.string.smartspace_icu_date_pattern_persian),
+                PersianDateFormat.PersianDateNumberCharacter.FARSI
+            )
+            this.formatterPersian = formatter
+        }
+        return formatterPersian!!.format(PersianDate(java.lang.Long.valueOf(System.currentTimeMillis())))
+    }
```
Comment (ghost) [inline lawnchair/src/app/lawnchair/smartspace/IcuDateTextView.kt]: I’d suggest avoiding the use of not-null assertion here.

```suggestion
    private fun getTimeTextPersian(updateFormatter: Boolean): String {
        val formatter = formatterPersian.takeIf { updateFormatter.not() } ?: PersianDateFormat(
            context.getString(R.string.smartspace_icu_date_pattern_persian),
            PersianDateFormat.PersianDateNumberCharacter.FARSI,
        ).also { formatterPersian = it }
        return formatter.format(PersianDate(java.lang.Long.valueOf(System.currentTimeMillis())))
    }
```

File: lawnchair/src/app/lawnchair/smartspace/IcuDateTextView.kt [hunk 4]
```diff
@@ -39,27 +51,62 @@ class IcuDateTextView @JvmOverloads constructor(
...
+    private fun getTimeTextGregorian(updateFormatter: Boolean): String {
+        if (formatterGregorian == null || updateFormatter) {
+            val formatter: DateFormat = DateFormat.getInstanceForSkeleton(
+                context.getString(R.string.smartspace_icu_date_pattern_gregorian),
+                Locale.getDefault()
+            )
+            this.formatterGregorian = formatter
+            formatter.setContext(DisplayContext.CAPITALIZATION_FOR_BEGINNING_OF_SENTENCE)
+        }
+        return formatterGregorian!!.format(java.lang.Long.valueOf(System.currentTimeMillis()))
```
Comment (ghost) [inline lawnchair/src/app/lawnchair/smartspace/IcuDateTextView.kt]: I’d suggest avoiding the use of not-null assertion here.

```suggestion
        val formatter = formatterGregorian.takeIf { updateFormatter.not() }
            ?: DateFormat.getInstanceForSkeleton(
                context.getString(R.string.smartspace_icu_date_pattern_gregorian),
                Locale.getDefault()
            ).also { formatterGregorian = it }
        return formatter.format(java.lang.Long.valueOf(System.currentTimeMillis()))
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: AndroidManifest-common.xml (hunk 2 above)
Comment (yasanglass) [reply AndroidManifest-common.xml:82]: It is. It basically tells gradle to use the `android:label` value from this manifest file and not from a conflicting one from the Persian Date library. Basically, it won't build with that dependency if this line is not added here.

File: lawnchair/src/app/lawnchair/smartspace/IcuDateTextView.kt (hunk 4 above)
Comment (yasanglass) [reply lawnchair/src/app/lawnchair/smartspace/IcuDateTextView.kt]: This is the exact original code and I just tried to not modify it and keep the same format even for the Persian Date function. It's now fixed on the new force push.

**Comment by yasanglass** [comment]
@patrykmichalik Should be ok now.

File: AndroidManifest-common.xml (hunk 2 above)
Comment (ghost) [reply AndroidManifest-common.xml:82]: Got it. Thanks!

File: lawnchair/src/app/lawnchair/smartspace/IcuDateTextView.kt (hunk 4 above)
Comment (ghost) [reply lawnchair/src/app/lawnchair/smartspace/IcuDateTextView.kt]: Ah, sorry, I didn’t notice that you’d used the preexisting code. If I had, I wouldn’t have requested this change, as it’s not entirely related to the pull request. I would’ve probably made it myself after the merge. But if you’ve already applied it, I don’t think it hurts to keep it. My apologies for the error.

**Review by ghost — APPROVED** [review]

**Pull request merged**